x-pack/filebeat/input/httpjson/input.go

```diff
@@ -81,532 +47,146 @@ func newRetryLogger() *retryLogger {
+		Name:       inputName,
+		Stability:  feature.Beta,
+		Deprecated: false,
+		Manager:    stateless.NewInputManager(configure),
 	}
-	return "", nil
 }
 
-// getRateLimit get the rate limit value if specified in the HTTP Header of the response,
-// and returns an init64 value in seconds since unix epoch for rate limit reset time.
-// When there is a remaining rate limit quota, or when the rate limit reset time has expired, it
-// returns 0 for the epoch value.
-func getRateLimit(header http.Header, rateLimit *RateLimit) (int64, error) {
-	if rateLimit != nil {
-		if rateLimit.Remaining != "" {
-			remaining := header.Get(rateLimit.Remaining)
-			if remaining == "" {
-				return 0, errors.Errorf("field %s does not exist in the HTTP Header, or is empty", rateLimit.Remaining)
-			}
-			m, err := strconv.ParseInt(remaining, 10, 64)
-			if err != nil {
-				return 0, errors.Wrapf(err, "failed to parse rate-limit remaining value")
-			}
-			if m == 0 {
-				reset := header.Get(rateLimit.Reset)
-				if reset == "" {
-					return 0, errors.Errorf("field %s does not exist in the HTTP Header, or is empty", rateLimit.Reset)
-				}
-				epoch, err := strconv.ParseInt(reset, 10, 64)
-				if err != nil {
-					return 0, errors.Wrapf(err, "failed to parse rate-limit reset value")
-				}
-				if time.Unix(epoch, 0).Sub(time.Now()) <= 0 {
-					return 0, nil
-				}
-				return epoch, nil
-			}
-		}
+func configure(cfg *common.Config) (stateless.Input, error) {
+	conf := defaultConfig()
+	if err := cfg.Unpack(&conf); err != nil {
+		return nil, err
 	}
-	return 0, nil
-}
 
-// applyRateLimit applies appropriate rate limit if specified in the HTTP Header of the response
-func (in *HttpjsonInput) applyRateLimit(ctx context.Context, header http.Header, rateLimit *RateLimit) error {
-	epoch, err := getRateLimit(header, rateLimit)
-	if err != nil {
-		return err
-	}
-	t := time.Unix(epoch, 0)
-	w := time.Until(t)
-	if epoch == 0 || w <= 0 {
-		in.log.Debugf("Rate Limit: No need to apply rate limit.")
-		return nil
-	}
-	in.log.Debugf("Rate Limit: Wait until %v for the rate limit to reset.", t)
-	ticker := time.NewTicker(w)
-	defer ticker.Stop()
-	select {
-	case <-ctx.Done():
-		in.log.Info("Context done.")
-		return nil
-	case <-ticker.C:
-		in.log.Debug("Rate Limit: time is up.")
-		return nil
-	}
+	return newHTTPJSONInput(conf)
 }
 
-// createRequestInfoFromBody creates a new RequestInfo for a new HTTP request in pagination based on HTTP response body
-func createRequestInfoFromBody(config *Pagination, response, last common.MapStr, ri *RequestInfo) (*RequestInfo, error) {
-	// we try to get it from last element, if not found, from the original response
-	v, err := last.GetValue(config.IDField)
-	if err == common.ErrKeyNotFound {
-		v, err = response.GetValue(config.IDField)
-	}
-
-	if err == common.ErrKeyNotFound {
-		return nil, nil
+func newHTTPJSONInput(config config) (*httpJSONInput, error) {
+	if err := config.Validate(); err != nil {
+		return nil, err
 	}
```

Comment:
note: Validate is already called by Unpack. But I like it to be called here again, as the constructor should be independent of `configure`.